binary_search_tree/tree.py

```diff
@@ -1,59 +1,144 @@
+        while current:
+            if current.key == key:
+                return current.value
+            elif key < current.key:  # if there is a left node keep searching
+                current = current.left
+            else:
+                current = current.right
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def inorder_helper(self, root, elements):
+        if root == None:
+            return
+
+        # if there is left, traverse
+        self.inorder_helper(root.left, elements)
+        # append the last left node that becomes the "root"
+        elements.append({"key": root.key, "value": root.value})
+        # if there is right, traverse (it will start from left recursive)
+        self.inorder_helper(root.right, elements)
+        return
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        # order: left, root, right
+        # if there are no nodes, return empty list
+        if self.root == None:
+            return []
 
-    # Time Complexity: 
-    # Space Complexity:     
+        elements = []
+        self.inorder_helper(self.root, elements)
+
+        return elements
+
+    def preorder_helper(self, root, elements):
+        if root == None:
+            return
+
+        # if there is root, add that
+        elements.append({"key": root.key, "value": root.value})
+        # if there is a left, traverse that
+        self.preorder_helper(root.left, elements)
+        # then traverse right
+        self.preorder_helper(root.right, elements)
+        return
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def preorder(self):
-        pass
+        # order: root, left, right
+        if self.root == None:
+            return []
+
+        elements = []
+        self.preorder_helper(self.root, elements)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, root, elements):
+        if root == None:
+            return
+
+        self.postorder_helper(root.left, elements)
+        self.postorder_helper(root.right, elements)
+        elements.append({"key": root.key, "value": root.value})
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def postorder(self):
-        pass
+        # order: left, right, current
+        if self.root == None:
+            return []
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def height(self):
-        pass
+        elements = []
+        self.postorder_helper(self.root, elements)
+
+    def max_height_helper(self, root, height):
+        if root == None:
+            return height
 
+        # else return +1 of left and right subtress
+        left_height = 1 + self.max_height_helper(root.left, height)
+        right_height = 1 + self.max_height_helper(root.right, height)
+
+        # compare to find max
+        if left_height < right_height:
+            return right_height
+        else:
+            return left_height
```

Comment:
You can also do this with the `max` function

```suggestion
        left_height = 1 + self.max_height_helper(root.left, height)
        right_height = 1 + self.max_height_helper(root.right, height)

        return max(right_height, left_height)
```